File: frontend/src/lib/courses.ts
import { apiFetch } from "./api";

export type Course = {
  id: number;
  owner_id: number;
  name: string;
  course_type: string;
  suitable_majors: string[];
  suitable_stages: string[];
  goal: string | null;
  price: string | number | null;
  teacher_name: string | null;
  description: string | null;
  created_at: string;
  updated_at: string;
};

export type CoursePayload = {
  name: string;
  course_type: string;
  suitable_majors: string[];
  suitable_stages: string[];
  goal: string | null;
  price: number | null;
  teacher_name: string | null;
  description: string | null;
};

export type CourseOptions = {
  course_types: string[];
  majors: string[];
  suitable_stages: string[];
};

export type CourseListResponse = {
  items: Course[];
  total: number;
};

export type CourseListQuery = {
  q?: string;
  course_type?: string;
  major?: string;
  stage?: string;
};

function toQuery(params: CourseListQuery): string {
  const search = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value === undefined || value === null || value === "") return;
    search.set(key, String(value));
  });
  const qs = search.toString();
  return qs ? `?${qs}` : "";
}

export function fetchCourseOptions() {
  return apiFetch<CourseOptions>("/api/courses/options");
}

export function fetchCourses(params: CourseListQuery = {}) {
  return apiFetch<CourseListResponse>(`/api/courses${toQuery(params)}`);
}

export function fetchCourse(id: number) {
  return apiFetch<Course>(`/api/courses/${id}`);
}

export function createCourse(payload: CoursePayload) {
  return apiFetch<Course>("/api/courses", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateCourse(id: number, payload: CoursePayload) {
  return apiFetch<Course>(`/api/courses/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export function deleteCourse(id: number) {
  return apiFetch<void>(`/api/courses/${id}`, { method: "DELETE" });
}

export function formatPrice(price: string | number | null) {
  if (price === null || price === undefined || price === "") return "未定价";
  const num = typeof price === "number" ? price : Number(price);
  if (Number.isNaN(num)) return String(price);
  return `¥${num.toLocaleString("zh-CN", { minimumFractionDigits: 0, maximumFractionDigits: 2 })}`;
}

File: backend/app/services/courses.py
from sqlalchemy.orm import Session

from app.models import Course
from app.schemas.course import CourseCreate, CourseUpdate


def list_courses(
    db: Session,
    owner_id: int,
    *,
    q: str | None = None,
    course_type: str | None = None,
    major: str | None = None,
    stage: str | None = None,
) -> list[Course]:
    query = db.query(Course).filter(Course.owner_id == owner_id)
    if q:
        like = f"%{q.strip()}%"
        query = query.filter(
            (Course.name.ilike(like))
            | (Course.teacher_name.ilike(like))
            | (Course.goal.ilike(like))
        )
    if course_type:
        query = query.filter(Course.course_type == course_type)

    items = query.order_by(Course.updated_at.desc(), Course.id.desc()).all()

    # JSON 字段筛选在应用层处理，兼容 SQLite
    if major:
        items = [item for item in items if major in (item.suitable_majors or [])]
    if stage:
        items = [item for item in items if stage in (item.suitable_stages or [])]
    return items


def get_owned_course(db: Session, course_id: int, owner_id: int) -> Course | None:
    return db.query(Course).filter(Course.id == course_id, Course.owner_id == owner_id).first()


def create_course(db: Session, owner_id: int, payload: CourseCreate) -> Course:
    course = Course(owner_id=owner_id, **payload.model_dump())
    db.add(course)
    db.commit()
    db.refresh(course)
    return course


def update_course(db: Session, course: Course, payload: CourseUpdate) -> Course:
    for key, value in payload.model_dump().items():
        setattr(course, key, value)
    db.add(course)
    db.commit()
    db.refresh(course)
    return course


def delete_course(db: Session, course: Course) -> None:
    db.delete(course)
    db.commit()
